src/peft/tuners/lora/torchao.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2024-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import warnings
+from typing import Any, Optional
+
+import torch
+
+# from torch import nn
+from peft.import_utils import is_torchao_available
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
+
+from .config import LoraConfig
+from .layer import Linear
+
+
+class TorchaoLoraLinear(Linear):
+    """LoRA layer implementation for Linear layers using torchao data"""
+
+    def __init__(self, *args, get_apply_tensor_subclass, **kwargs):
+        # this is not strictly necessary, as kwargs are stored either way, but we want to error early if
+        # get_apply_tensor_subclass is missing.
+        super().__init__(*args, **kwargs)
+        self.get_apply_tensor_subclass = get_apply_tensor_subclass
+        self._check_dtype_supported()
+
+    def _check_dtype_supported(self):
+        # TODO: Not required once int4_weight_only is properly supported by torchao
+        base_layer = self.get_base_layer()
+        weight = base_layer.weight
+        if hasattr(weight, "layout_tensor") and (weight.layout_tensor.data.dtype != torch.int8):
+            raise ValueError(f"{type(self).__name__} only supports int8 weights for now.")
+
+    def merge(self, safe_merge: bool = False, adapter_names: Optional[list[str]] = None) -> None:
+        from torchao import quantize_
```

Comment:
`quantize_` is only available from torchao 0.4.0. Maybe we should modify a bit `is_torchao_available` to take that into account ?